### PR TITLE
Add synthetic weight support to DeepSeek V3 MLA tests

### PR DESCRIPTION
**Pull request by ntarafdar**
## Summary

This PR adds support for synthetic (random) weights to the DeepSeek V3 MLA tests, enabling testing without requiring downloaded model weights. This is useful for:
- Development and debugging when model weights are unavailable
- Rapid iteration on test infrastructure
- Verifying numerical stability with controlled inputs
- CI/CD pipelines where downloading large models is impractical

### 1. Added Synthetic Weight Generation
- Added `generate_synthetic_state_dict()` function that creates properly formatted synthetic weights
- Generates reproducible random weights using a fixed seed (42 + layer_idx)
- Matches exact DeepSeek V3 MLA architecture dimensions

### 2. Environment Variable Control
- Added `USE_SYNTHETIC_WEIGHTS` environment variable flag to enable synthetic weight mode
- When set to "1", tests use synthetic weights instead of downloaded weights
- Default behavior (when unset or "0") continues to use downloaded weights

### 3. Modified Reference I/O Generation
- Updated `generate_reference_io()` to support both downloaded and synthetic weight paths
- Ensures identical synthetic weights are used for both reference (PyTorch) and device (TTNN) computations
- Bypasses quantization for synthetic weights while maintaining correct tensor formats

### 4. Weight Format Handling
- Converts synthetic weights to float8_e4m3fn format expected by TTNN
- Properly handles scale tensors with block-based dimensions (128x128 blocks)
- Uses math.ceil for non-aligned weight dimensions to ensure proper coverage

### Weight Generation
The synthetic weights are generated with the following characteristics:
- **Reproducible**: Uses fixed seed for consistent results across runs
- **Architecture-Aligned**: Generates weights matching exact MLA dimensions:
  - `q_a_proj`: (q_lora_rank, hidden_size)
  - `q_b_proj`: (q_head_dim * n_heads, q_lora_rank)  
  - `kv_a_proj_with_mqa`: (kv_lora_rank + qk_rope_head_dim, hidden_size)
  - `kv_b_proj`: (kv_head_dim * n_kv_heads, kv_lora_rank)
  - `o_proj`: (hidden_size, v_head_dim * n_heads)

### Format Conversion
```python
# Convert to float8_e4m3fn format that TTNN expects
state_dict[name] = clamped_tensor.to(torch.float8_e4m3fn)

# Calculate scale shape - round up to ensure coverage
import math
scale_shape = [math.ceil(tensor.shape[i] / block_size[i]) for i in range(len(tensor.shape))]
state_dict[name + "_scale_inv"] = torch.ones(scale_shape, dtype=torch.float32)
```

### Manual Testing

1. **Set the environment variable to enable synthetic weights:**
```bash
export USE_SYNTHETIC_WEIGHTS=1
```

2. **Set up the device configuration:**
```bash
export MESH_DEVICE=TG  # or T3K depending on your hardware
export DEEPSEEK_V3_CACHE="/tmp/deepseek-v3-cache-synthetic"  # Use temp directory
mkdir -p "$DEEPSEEK_V3_CACHE"
```

3. **Run a specific decode test with synthetic weights:**
```bash
pytest -xvs models/demos/deepseek_v3/tests/test_mla.py::test_forward_pass[run_test_forward_pass_mla1d-None-device_params0-mode_decode_seq_1_batch_32_pos_random] --no-header --tb=short
```

4. **To disable synthetic weights and return to downloaded weights:**
```bash
unset USE_SYNTHETIC_WEIGHTS
```

## Test Validation Results

The synthetic weight path has been validated with successful test runs achieving high PCC values:
- **Output PCC**: 0.9945 (requirement: >= 0.99) ✅
- **Cache KV PCC**: 0.9992 (requirement: >= 0.999) ✅
- **Cache PE PCC**: 0.9994 (requirement: >= 0.999) ✅

These results confirm that the synthetic weights maintain numerical stability and produce consistent outputs between reference and device implementations.

## Benefits

1. **No External Dependencies**: Tests can run without downloading large model files
2. **Faster CI/CD**: Eliminates model download time in automated pipelines
3. **Reproducible Testing**: Fixed seed ensures consistent test behavior
4. **Development Flexibility**: Enables testing when model access is restricted
5. **Backward Compatible**: All existing tests continue to work unchanged

## Testing

- All existing tests continue to work with downloaded weights (default behavior)
- The synthetic weight path is only activated when explicitly enabled via the environment variable
- No changes to test signatures or existing test infrastructure
- Fully backward compatible with existing test workflows

## Future Improvements

Potential enhancements for future PRs:
- Add pytest command-line option as an alternative to environment variable
- Support for different random distributions for weight initialization
- Configuration file support for synthetic weight parameters
- Integration with CI/CD pipeline configuration

### CODE REVIEW

**Comment by yieldthought** [comment]
Codex review:

[models/demos/deepseek_v3/tests/test_mla.py:256](https://github.com/tenstorrent/tt-metal/blob/64d384dc8cfda82cbc53801b01849b11fed777b3/models/demos/deepseek_v3/tests/test_mla.py#L256) - In synthetic mode the module_path branch returns a state_dict with keys like model.layers.0.self_attn.q_a_proj.weight, but the test still passes that dict directly to MLA1D/MLA2D.convert_weights, whose converters request q_a_proj.weight, kv_b_proj.weight_scale_inv, etc. The non-synthetic branch strips the prefix before returning. Since test_forward_pass parameterizes module_path with "model.layers.0.self_attn", USE_SYNTHETIC_WEIGHTS=1 will fail those cases during weight conversion before exercising the synthetic path. Please keep the returned synthetic state_dict module-local, or strip the prefix before returning, for this direct MLA test.

[models/demos/deepseek_v3/tests/test_mla.py:433](https://github.com/tenstorrent/tt-metal/blob/64d384dc8cfda82cbc53801b01849b11fed777b3/models/demos/deepseek_v3/tests/test_mla.py#L433) - The synthetic flag changes the actual state_dict used for conversion, but the weight-cache path is still exactly the same pytest-node cache used by downloaded weights; get_test_weight_config only namespaces by PYTEST_CURRENT_TEST/layers/mesh. With --reset-weight-cache omitted, a previous real-weight run can be reused for USE_SYNTHETIC_WEIGHTS=1, and a synthetic run can pollute subsequent real-weight runs. Please include the weight source in the cache path or force recalculation when synthetic weights are enabled. The same issue applies to the MLA2D call at [models/demos/deepseek_v3/tests/test_mla.py:576](https://github.com/tenstorrent/tt-metal/blob/64d384dc8cfda82cbc53801b01849b11fed777b3/models/demos/deepseek_v3/tests/test_mla.py#L576).

**Review by yieldthought — APPROVED** [review]
Approved to unblock but be careful here; see codex comments about cross-cache pollution >.<

**Comment by tdowdallTT** [comment]
**AUTOMATION**: Closing this PR because it has had no human activity for 132 days. Automated updates (priority/complexity labelling, CODEOWNERS re-labelling and bot comments) are not counted as activity, which is why it stayed open past the usual 32-day window.

If you still need this change, please reopen it or open a fresh PR whenever you're ready to continue.
